README.md
[![Build Status](https://travis-ci.org/fkooman/php-lib-http.png?branch=master)](https://travis-ci.org/fkooman/php-lib-http)
[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/fkooman/php-lib-http/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/fkooman/php-lib-http/?branch=master)

# Introduction

Simple HTTP helper library.

# Features

This library makes it easy to:

- handle HTTP requests;
- create and send HTTP responses;
- handle sessions

# API

## Request

Typically you'd use the `Request` class to construct an object based on the 
current request:

    use fkooman\Http\Request;

    // creates the object from $_SERVER, $_POST or php://input
    $r = new Request();
    
    // shows the request method, e.g. GET/POST/DELETE
    echo $r->getMethod();

    // shows the full request URL
    echo $r->getUrl();

    // get POST parameter
    echo $r->getPostParameter('foo');

    // get HTTP header value
    echo $r->getHeader('Accept');

## Response

    use fkooman\Http\Response;

    // create the object with response code and content type
    $r = new Response(200, 'text/plain');

    // set a response header
    $r->setHeader('X-Foo', 'Bar');

    // set the body content
    $r->setBody('Simple Plain Text');

    // send the response
    $r->send();

There is also an `fkooman\Http\JsonResponse` for sending JSON responses, the
`setBody()` method accepts an `array`. 

For sending redirects to the browser `fkooman\Http\RedirectResponse` is 
available:

    use fkooman\Http\RedirectResponse;

    // send a (temporary) redirect
    $r = new RedirectResponse('https://www.example.org/', 302);
    $r->send();

## Session

    use fkooman\Http\Session;

    // create the session
    $s = new Session('My Session');

    // set, get, delete variables
    $s->set('foo', 'bar');
    $s->get('foo');
    $s->delete('foo');

    // check whether a variable is set
    $s->has('foo');

    // destroy a session
    $s->destroy();

# Installation

You can use this library through [Composer](http://getcomposer.org/) by 
requiring `fkooman/http`:

    $ composer require fkooman/http

# Deployment Considerations

## Apache

    ServerName https://server.name:443
    UseCanonicalName on

See https://httpd.apache.org/docs/2.4/mod/core.html for `ServerName`, 
`UseCanonicalName` and `UseCanonicalPhysicalPort`.

Without `UseCanonicalName on` the `SERVER_NAME` variable is set using the 
`HTTP_HOST` header. This is correct in most cases, but if you want to use 
the `SERVER_NAME` as something reliable under your web server's control you 
need to set `UseCanonicalName on`.

# Tests

Extensive tests for PHPUnit are available.

# License

Licensed under the Apache License, Version 2.0;

   http://www.apache.org/licenses/LICENSE-2.0


CHANGES.md
# Release History

## 2.0.0 (2016-08-19)
- remove `FormResponse`
- refactor `Url` class to use different HTTP server headers to determine the 
  exact URL, remove a lot of cruft
- require PHP >= 5.4 now
- use mb functions wherever possible

## 1.7.0 (2016-06-07)
- remove `fkooman/json` dependency, directly use `json_encode` and 
  `json_decode`

## 1.6.2 (2016-03-25)
- try to fix incompatible types for travis-ci

## 1.6.1 (2016-03-25)
- update `fkooman/json`
- missing `use Exception` from exceptions
- small documentation updates

## 1.6.0 (2016-02-22)
- add `getBody()` to `Response` class

## 1.5.1 (2015-12-22)
- default `RedirectResponse` status code is `302` now, `301` is dangerous
  and should only EXPLICITLY be used

## 1.5.0 (2015-12-21)
- prefer `HTTP_HOST` in requests to determine the request URL instead of 
  relying on `SERVER_NAME` (`SERVER_NAME` will be removed in 2.0.0)
- **DEPRECATE** `Url::getPort()`, applications SHOULD not use this (
  `Url::getPort()` will be removed in 2.0.0)

## 1.4.0 (2015-12-14)
- implement Response::getStatusCode() and Response::isOkay()

## 1.3.2 (2015-12-11)
- fix running behind a HTTP proxy

## 1.3.1 (2015-11-23)
- allow the `description` parameter from `UnauthorizedException` to be
  missing
- make `QUERY_STRING` optional in `Uri` class as it is not set by default by 
  PHP built in web server

## 1.3.0 (2015-11-17)
- make `Session` implement `SessionInterface` so applications can use
  that to make testing much easier

## 1.2.0 (2015-11-10)
- add `setFile($fileName)` method to `Response` class which sets the
  `X-SendFile` header (for mod_xsendfile)

## 1.1.3 (2015-11-01)
- undo changes of 1.1.2, needs furter investigation to be sure we are
  taking the right approach.

## 1.1.2 (2015-10-30)
- use `rawurldecode` on `PATH_INFO` to match behavior when using php-fpm 
  instead of mod_php.

## 1.1.1 (2015-10-13)
- allow providing raw request body to `Request` object

## 1.1.0 (2015-10-07)
- add `Content-Length` header to responses

## 1.0.1 (2015-09-07)
- fix running the unit test on CentOS 6

## 1.0.0 (2015-09-07)
- initial release


src/fkooman/Http/Url.php
<?php

/**
 * Copyright 2016 François Kooman <fkooman@tuxed.net>.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
namespace fkooman\Http;

use RuntimeException;

/**
 * Url Class.
 *
 * Helper class to easily determine the request URL. It has a number of
 * important functions:
 *
 * - Determine the REST route (i.e.: PATH_INFO) that should be executed. There
 *   are a number of scenarios we want to support. For example of the requested
 *   route is /foo/bar the request could look like any of these:
 *   - http://www.example.org/app/index.php/foo/bar (in folder 'app'
 *   - http://www.example.org/app/foo/bar (URL rewriting to index.php)
 *   - http://www.example.org/index.php/foo/bar (in the root)
 *   - http://www.example.org/foo/bar (in the root with rewriting)
 *
 *   We want to make it possible to figure out the (relative and absolute) path
 *   of the application, if it runs in the root we want to return '/', if it
 *   runs in a folder we want to return /app/.
 *
 *   In essence, we need to detect whether or not URL rewriting is active and
 *   whether or not the application is located in the root or in a folder.
 *   Various server variables give different results for different web servers
 *   and configurations and also for different PHP versions and whether or not
 *   php-fpm is used. This all needs to be made consisent.
 *
 * - Make the query parameters easily available.
 *
 * - Make it possible to reconstruct the full request URL from the server
 *   variables.
 */
class Url
{
    /** @var string */
    private $scheme;

    /** @var string */
    private $serverName;

    /** @var int */
    private $serverPort;

    /** @var string */
    private $requestUri;

    /** @var string */
    private $pathInfo;

    /**
     * Create the Url object.
     *
     * @param array $srv the server variables, typically $_SERVER
     */
    public function __construct(array $srv)
    {
        $this->scheme = self::extractScheme($srv);

        $requiredKeys = [
            'SERVER_NAME',      // e.g. foo.example.org
            'SERVER_PORT',      // e.g. 443
            'REQUEST_URI',      // e.g. /foo/bar?baz=123
        ];
        foreach ($requiredKeys as $k) {
            if (!array_key_exists($k, $srv)) {
                throw new RuntimeException(sprintf('missing key "%s"', $k));
            }
        }

        $this->pathInfo = self::extractPathInfo($srv);

        $this->serverName = $srv['SERVER_NAME'];
        $this->serverPort = intval($srv['SERVER_PORT']);
        $this->requestUri = $srv['REQUEST_URI'];
    }

    /**
     * Get the URL scheme.
     *
     * @return string the URL scheme
     */
    public function getScheme()
    {
        return $this->scheme;
    }

    /**
     * Get the URL host.
     *
     * @return string the URL host
     */
    public function getHost()
    {
        return $this->serverName;
    }

    /**
     * Get the URL port.
     *
     * @return int the URL port
     */
    public function getPort()
    {
        return $this->serverPort;
    }

    /**
     * Get the REQUEST_URI.
     *
     * @return string the REQUEST_URI
     */
    public function getRequestUri()
    {
        return $this->requestUri;
    }

    /**
     * Get the PATH_INFO.
     *
     * @return string the PATH_INFO
     */
    public function getPathInfo()
    {
        return $this->pathInfo;
    }

    /**
     * The query string.
     *
     * @return string the query string, or empty string if no query string
     *                is available
     */
    public function getQueryString()
    {
        $requestUri = $this->getRequestUri();

        if (false !== $qPos = mb_strpos($requestUri, '?')) {
            // has query string
            return mb_substr($requestUri, $qPos + 1);
        }

        return '';
    }

    /**
     * The query string as array.
     *
     * @return array the query string as array. The array will be empty if
     *               the query string is empty
     */
    public function getQueryStringAsArray()
    {
        $queryString = $this->getQueryString();

        if ('' === $queryString) {
            return [];
        }

        $queryStringArray = [];
        parse_str($queryString, $queryStringArray);

        return $queryStringArray;
    }

    /**
     * Return a specific query string parameter value.
     *
     * @param string $key the query parameter key to get
     *
     * @return mixed the query parameter value if it is set, or null if the
     *               parameter is not available
     */
    public function getQueryParameter($key)
    {
        $queryStringArray = $this->getQueryStringAsArray();

        if (array_key_exists($key, $queryStringArray)) {
            return $queryStringArray[$key];
        }

        return null;
    }

    /**
     * Get the REQUEST_URI without PATH_INFO and QUERY_STRING, taking server
     * rewriting in consideration.
     *
     * Example (without URL rewriting):
     * https://www.example.org/foo/index.php/bar?a=b will return:
     * '/foo/index.php'
     *
     * Example (with URL rewriting to index.php):
     * https://www.example.org/foo/bar?a=b will return:
     * '/foo'
     *
     * Example (with URL rewriting to index.php without sub folder):
     * https://www.example.org/bar?a=b will return:
     * ''
     */
    public function getRoot()
    {
        $requestUri = $this->getRequestUri();
        $pathInfo = $this->getPathInfo();

        // remove query string from REQUEST_URI, if set
        if (false !== $qPos = mb_strpos($requestUri, '?')) {
            // has query string
            $requestUri = mb_substr($requestUri, 0, $qPos + 1);
        }

        // remove PATH_INFO from REQUEST_URI, if set
        if (false !== $pPos = mb_strrpos($requestUri, $pathInfo)) {
            // has PATH_INFO
            $requestUri = mb_substr($requestUri, 0, $pPos + 1);
        }

        return $requestUri;
    }

    /**
     * Get the root as a full URL.
     */
    public function getRootUrl()
    {
        return $this->getAuthority().$this->getRoot();
    }

    /**
     * Get the authority part of the URL. That is, the scheme, host and
     * optional port if it is not a standard port.
     *
     * @return string the authority part of the URL
     */
    public function getAuthority()
    {
        $scheme = $this->getScheme();
        $serverName = $this->getHost();
        $serverPort = $this->getPort();

        $authority = sprintf('%s://%s', $scheme, $serverName);
        if (('https' === $scheme && 443 !== $serverPort) || ('http' === $scheme && 80 !== $serverPort)) {
            $authority .= sprintf(':%d', $serverPort);
        }

        return $authority;
    }

    /**
     * Get the URL as a string.
     */
    public function toString()
    {
        return $this->getAuthority().$this->getRequestUri();
    }

    /**
     * Get the URL as a string if it is coerced to string.
     */
    public function __toString()
    {
        return $this->toString();
    }

    private static function extractScheme(array $srv)
    {
        // prefer REQUEST_SCHEME variable if set
        if (array_key_exists('REQUEST_SCHEME', $srv)) {
            return $srv['REQUEST_SCHEME'];
        }

        // fallback to HTTPS variable
        if (array_key_exists('HTTPS', $srv)) {
            if ('' !== $srv['HTTPS'] && 'off' !== $srv['HTTPS']) {
                return 'https';
            }

            return 'http';
        }

        // default to "http" if we cannot find anything
        return 'http';
    }

    private static function extractPathInfo(array $srv)
    {
        if (array_key_exists('PATH_INFO', $srv)) {
            return $srv['PATH_INFO'];
        }

        // On CentOS 7 with PHP 5.4 PATH_INFO is null when rewriting is
        // enabled and you go to the root. On Fedora 22 with PHP 5.6
        // PATH_INFO is '/' in the same scenario.
        return '/';
    }
}
